cli/command/manifest/push.go

```diff
@@ -0,0 +1,591 @@
+			} else if ips, err := net.LookupHost(reg); err == nil {
+				insecureRegistries = append(insecureRegistries, ips...)
+			} else {
+				return []string{}, errors.Wrapf(err, "manifest create: Invalid registry (%s) specified in ~/.docker/config.json: %s", reg)
+			}
+		}
+	}
+
+	return insecureRegistries, nil
+}
+
+func pushReferences(httpClient *http.Client, urlBuilder *v2.URLBuilder, ref reference.Named, manifests []manifestPush) error {
+	for _, manifest := range manifests {
+		dgst, err := digest.Parse(manifest.Digest)
+		logrus.Debugf("pushing ref digest %s", dgst)
+		if err != nil {
+			return errors.Wrapf(err, "error parsing manifest digest (%s) for referenced manifest %q: %v", manifest.Digest, manifest.Name)
+		}
+		targetRef, err := reference.WithDigest(ref, dgst)
+		logrus.Debugf("pushing ref %v", targetRef)
+		if err != nil {
+			return errors.Wrapf(err, "error creating manifest digest target for referenced manifest %q: %v", manifest.Name)
+		}
+		pushURL, err := urlBuilder.BuildManifestURL(targetRef)
+		if err != nil {
+			return errors.Wrapf(err, "error setting up manifest push URL for manifest references for %q: %v", manifest.Name)
+		}
+		logrus.Debugf("manifest reference push URL: %s", pushURL)
+
+		pushRequest, err := http.NewRequest("PUT", pushURL, bytes.NewReader(manifest.JSONBytes))
+		if err != nil {
+			return errors.Wrap(err, "HTTP PUT request creation for manifest reference push failed")
+		}
+		pushRequest.Header.Set("Content-Type", manifest.MediaType)
+		resp, err := httpClient.Do(pushRequest)
+		if err != nil {
+			return errors.Wrap(err, "PUT of manifest reference failed")
+		}
+
+		resp.Body.Close()
+		if !statusSuccess(resp.StatusCode) {
+			return fmt.Errorf("referenced manifest push unsuccessful: response %d: %s", resp.StatusCode, resp.Status)
+		}
+		dgstHeader := resp.Header.Get("Docker-Content-Digest")
+		dgstResult, err := digest.Parse(dgstHeader)
+		if err != nil {
+			return errors.Wrap(err, "couldn't parse pushed manifest digest response")
+		}
+		if string(dgstResult) != manifest.Digest {
+			return fmt.Errorf("pushed referenced manifest received a different digest: expected %s, got %s", manifest.Digest, string(dgst))
+		}
+		logrus.Debugf("referenced manifest %q pushed; digest matches: %s", manifest.Name, string(dgst))
+	}
+	return nil
+}
+
+func mountBlobs(httpClient *http.Client, urlBuilder *v2.URLBuilder, ref reference.Named, blobsRequested []blobMount) error {
+
+	for _, blob := range blobsRequested {
+		// create URL request
+		url, err := urlBuilder.BuildBlobUploadURL(ref, url.Values{"from": {blob.FromRepo}, "mount": {blob.Digest}})
+		if err != nil {
+			return errors.Wrap(err, "failed to create blob mount URL")
+		}
+		mountRequest, err := http.NewRequest("POST", url, nil)
+		if err != nil {
+			return errors.Wrap(err, "HTTP POST request creation for blob mount failed")
+		}
+		mountRequest.Header.Set("Content-Length", "0")
+		resp, err := httpClient.Do(mountRequest)
```

Comment:
A possibly simpler way to do this is to use the registry client library and do the following:

```go
bs := repo.Blobs(ctx)
lu, err := bs.Create(ctx, client.WithMountFrom(fromRepo))
if err == distribution.ErrBlobMounted {
        // mount successful
} else if err == nil {
        // registry treated this as a normal upload
        lu.Cancel()
}
```